src/http/content.cr

```diff
@@ -47,106 +49,142 @@ module HTTP
+
+      byte = @io.read_byte
+      if byte
+        chunk_bytes_read 1
         byte
       else
-        super
+        raise IO::EOFError.new("Invalid HTTP chunked content: unexpected end of file")
       end
     end
 
     def peek
-      while true
-        if @chunk_remaining > 0
-          peek = @io.peek
-          return nil unless peek
-
-          if @chunk_remaining < peek.size
-            peek = peek[0, @chunk_remaining]
-          end
+      next_chunk
+      return nil if @chunk_remaining == 0
 
-          return peek
-        elsif @read_chunk_start
-          read_chunk_start
-          next
-        end
+      peek = @io.peek || return
 
-        break
+      if @chunk_remaining < peek.size
+        peek = peek[0, @chunk_remaining]
+      elsif peek.size == 0
+        raise IO::EOFError.new("Invalid HTTP chunked content: unexpected end of file")
       end
 
-      nil
+      peek
     end
 
     def skip(bytes_count)
       if bytes_count <= @chunk_remaining
         @io.skip(bytes_count)
-        @chunk_remaining -= bytes_count
-        check_chunk_remaining_is_zero
+        chunk_bytes_read bytes_count
       else
         super
       end
     end
 
-    private def check_chunk_remaining_is_zero
+    private def chunk_bytes_read(size)
+      @chunk_remaining -= size
+
       # As soon as we finish reading a chunk we return,
       # in case the next content is delayed (see #3270).
-      # We set @read_chunk_start to true so we read the next
+      # We set @expect_chunk_start to true so we read the next
       # chunk start on the next call to `read`.
       if @chunk_remaining == 0
-        @read_chunk_start = true
+        read_crlf
+        @expect_chunk_start = true
       end
     end
 
-    private def read_chunk_start
-      read_chunk_end
-      @chunk_remaining = @io.gets.not_nil!.to_i(16)
-      check_last_chunk
-      @read_chunk_start = false
+    # Check if the last read consumed a chunk and we
+    # need to start consuming the next one.
+    private def next_chunk
+      return unless @expect_chunk_start
+
+      if read_chunk_size == 0
+        read_trailer
+      end
+
+      @expect_chunk_start = false
     end
 
-    private def read_chunk_end
-      # Read "\r\n"
-      @io.skip(2)
+    private def read_crlf
+      bytes = Bytes.new(2)
+      @io.read_fully(bytes)
+      unless bytes == "\r\n".to_slice
+        raise IO::Error.new("Invalid HTTP chunked content: expected CRLF")
+      end
     end
 
-    private def check_last_chunk
-      # If we read "0\r\n", we need to read another "\r\n"
-      read_chunk_end if @chunk_remaining == 0
+    private def read_chunk_size
+      line = read_delimited_line
+
+      if index = line.byte_index(';'.ord)
+        chunk_size = line.byte_slice(0, index)
+      else
+        chunk_size = line
+      end
+
+      @chunk_remaining = chunk_size.to_i?(16) || raise IO::Error.new("Invalid HTTP chunked content: invalid chunk size (#{chunk_size.dump})")
```

Comment:
Let's avoid allocating a string: "invalid chunk size" is enough, not need to dump the line.

Since we ignore the optional headers, and those are fairly rare, I wonder whether we really have to seek for a potential `;`. Maybe that would suffice:

```crystal
chunk_size.to_i(16, strict: false) do
   raise IO::Error.new
end
```

Reply:
Yeah, it's not really commonly used. But I'd leave it in to be RFC compliant. It's implemented anyway.

But I wondered whether string allocation can be avoided... maybe if we could read an int from an IO? This could be useful in other cases as well.
A hex encoded Int32 has 8 digits max (+ sign). So it could be stored in a `UInt8[9]`...